server/sonar-server/src/main/java/org/sonar/server/computation/period/Period.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.period;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Calendar;
+import java.util.Date;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.builder.ReflectionToStringBuilder;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.utils.DateUtils;
+import org.sonar.core.component.SnapshotDto;
+
+public class Period {
+
+  private int index;
+  private String mode, modeParameter;
+  private SnapshotDto projectSnapshot;
+  /**
+   * Date resolved from the settings of the period
+   * - date : the given date
+   * - days : nearest analysis date found
+   * - previous analysis : date of the last analysis
+   * - previous version : date of the analysis of the previous version
+   * - version : date of the analysis of the given version
+   */
+  private Long targetDate = null;
+
+  public Period(String mode, @Nullable Long targetDate, SnapshotDto projectSnapshot) {
+    this.mode = mode;
+    if (targetDate != null) {
+      this.targetDate = targetDate;
+    }
+    this.projectSnapshot = projectSnapshot;
+  }
+
+  public Period setIndex(int index) {
+    this.index = index;
+    return this;
+  }
+
+  public int getIndex() {
+    return index;
+  }
+
+  public SnapshotDto getProjectSnapshot() {
+    return projectSnapshot;
+  }
+
+  /**
+   * Date of the snapshot
+   */
+  public Long getSnapshotDate() {
+    return projectSnapshot != null ? projectSnapshot.getCreatedAt() : null;
+  }
+
+  public String getMode() {
+    return mode;
+  }
+
+  public String getModeParameter() {
+    return modeParameter;
+  }
+
+  public Period setModeParameter(String s) {
+    this.modeParameter = s;
+    return this;
+  }
+
+  @VisibleForTesting
+  Long getTargetDate() {
+    return targetDate;
+  }
+
+  @Override
+  public String toString() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "toString" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)